packages/shopping-cart/src/use-shopping-cart.ts

```diff
@@ -1,11 +1,38 @@
-import { useContext } from 'react';
-import ShoppingCartContext from './shopping-cart-context';
-import type { ShoppingCartManager } from './types';
-
-export default function useShoppingCart(): ShoppingCartManager {
-	const manager = useContext( ShoppingCartContext );
-	if ( ! manager ) {
-		throw new Error( 'useShoppingCart must be used inside a ShoppingCartProvider' );
-	}
-	return manager;
+import { useContext, useState, useEffect, useRef } from 'react';
+import { createUseShoppingCartState } from './shopping-cart-hook-manager';
+import ShoppingCartOptionsContext from './shopping-cart-options-context';
+import useManagerClient from './use-manager-client';
+import useRefetchOnFocus from './use-refetch-on-focus';
+import type { UseShoppingCart } from './types';
+
+export default function useShoppingCart( cartKey?: string ): UseShoppingCart {
+	const managerClient = useManagerClient( 'useShoppingCart' );
+
+	const { defaultCartKey } = useContext( ShoppingCartOptionsContext );
+	const finalCartKey = cartKey ?? defaultCartKey;
+	const manager = managerClient.forCartKey( finalCartKey );
+
+	useRefetchOnFocus( finalCartKey );
+
+	useEffect( () => {
+		manager.fetchInitialCart();
+	}, [ manager ] );
+
+	const isMounted = useRef( true );
+	useEffect( () => {
+		isMounted.current = true;
+		return () => {
+			isMounted.current = false;
+		};
+	}, [] );
+
+	// Re-render when the cart changes
+	const [ cartState, setCartState ] = useState( createUseShoppingCartState( manager ) );
+	useEffect( () => {
+		return manager.subscribe( () => {
+			isMounted.current && setCartState( createUseShoppingCartState( manager ) );
+		} );
+	}, [ manager ] );
```

Comment:
This entire `isMounted` and `cartState` machinery could be replaced by the [`useSubscription`](https://github.com/facebook/react/tree/main/packages/use-subscription) hook:
```js
const cartSubscription = useMemo(
  () => ( {
    getCurrentValue: () => createUseShoppingCartState( manager ),
    subscribe: manager.subscribe,
  } ),
  [ manager ]
);

return useSubscription( cartSubscription );
```
It has the `isMounted` check built in, and covers many more edge cases. Like one where the `manager` fires an update between the initial `useState` reading state from `manager`, and the `useEffect` callback that creates the subscription to `manager`. Our simple implementation would miss such an update.